certora/specs/Marketplace.spec

```diff
@@ -341,6 +358,25 @@ rule functionsCausingRequestStateChanges(env e, method f) {
     assert requestStateBefore == Marketplace.RequestState.Started && requestStateAfter == Marketplace.RequestState.Failed => canFailRequest(f);
 }
 
+rule functionsCausingSlotStateChanges(env e, method f) {
+    calldataarg args;
+    Marketplace.SlotId slotId;
+
+    slotAttributesAreConsistent(e, slotId);
+
+    Marketplace.Slot slot = currentContract.slots(e, slotId);
+    Marketplace.SlotState slotStateBefore = currentContract.slotState(e, slotId);
+    f(e, args);
+    Marketplace.SlotState slotStateAfter = currentContract.slotState(e, slotId);
+
+    // SlotState.Finished -> SlotState.Paid
+    assert slotStateBefore == Marketplace.SlotState.Finished && slotStateAfter == Marketplace.SlotState.Paid => canMakeSlotPaid(f);
+
+    // SlotState.Free -> SlotState.Filled
+    assert slotStateBefore != Marketplace.SlotState.Filled && slotStateAfter == Marketplace.SlotState.Filled => canFillSlot(f);
+    assert slotStateBefore != Marketplace.SlotState.Filled && slotStateAfter == Marketplace.SlotState.Filled => slotStateBefore == Marketplace.SlotState.Free;
```

Comment:
Do you think it makes sense to have an `allowedSlotStateChanges` rule as well? Right now it seems it's doing a mix of "allowedSlotStateChanges" and "functionsCausingSlotStateChanges".

Not something that has to be done now, but I see some similarities to the request state change rules.

Reply:
Also, what about any `SlotState.Cancelled` variations?